include/cuco/detail/utils.cuh

```diff
@@ -92,12 +103,33 @@ struct slot_is_filled {
 template <typename SizeType, typename HashType>
 __host__ __device__ constexpr SizeType sanitize_hash(HashType hash) noexcept
 {
-  if constexpr (cuda::std::is_signed_v<SizeType>) {
-    return cuda::std::abs(static_cast<SizeType>(hash));
+  if constexpr (cuda::std::is_same_v<HashType, cuda::std::array<std::uint64_t, 2>>) {
+#if !defined(CUCO_HAS_INT128)
+    static_assert(false,
+                  "CUCO_HAS_INT128 undefined. Need unsigned __int128 type when sanitizing "
+                  "cuda::std::array<std::uint64_t, 2>");
+#endif
+    unsigned __int128 ret{};
+    memcpy(&ret, &hash, sizeof(unsigned __int128));
+    return to_positive<SizeType>(static_cast<SizeType>(ret));
   } else {
-    return static_cast<SizeType>(hash);
+    return to_positive<SizeType>(hash);
   }
 }
 
+/**
+ * @brief Converts a given hash value and cg_rank, into a valid (positive) size type.
+ *
+ * @tparam SizeType The target type
+ * @tparam HashType The input type
+ *
+ * @return Converted hash value
+ */
+template <typename SizeType, typename HashType>
+__host__ __device__ constexpr SizeType sanitize_hash(HashType hash, std::uint32_t cg_rank) noexcept
+{
+  return sanitize_hash<SizeType>(sanitize_hash<SizeType>(hash) + cg_rank);
```

Comment:
There exists a scenario where this approach fails and it's the reason CI in my initial draft PR failed.
Consider the following example:

- `SizeType` is `int32_t` aka a signed type
- `HashType` is `uint64_t` or any type that has `numeric_limits<HashType>::max() > numeric_limits<SizeType>::max()`
- The value of `sanitize_hash<SizeType>(hash)` is very close to `numeric_limits<SizeType>::max()`

In this scenario, if we compute `sanitize_hash<SizeType>(hash) + cg_rank` there's chance the result oxceeds `numeric_limits<SizeType>::max()` which would result in a signed integer overflow which is undefined behavior under the C++ abstract machine. Thus the compiler is free to produce any garbage code around this call.

To solve this we need check if `sanitize_hash<SizeType>(hash) > (numeric_limits<SizeType>::max() - group.size())` (be careful with `>` and `>=`, I'm infamous for my off-by-one errors) and then compute the wrapped-around value manually in case this expression evaluates to `true`

Reply:
Thanks for the detailed explanation! 
I think [3ae1afe](https://github.com/NVIDIA/cuCollections/pull/501/commits/3ae1afeb0e2b55d1b2ede3496c5de82a74541ce0) covers this case now.